### ISSUE TITLE
[java] UnusedLocalVariable false positive with annotated method ref

### ISSUE THREAD

**Issue by LynnBroe** [pmd/pmd#4687]
**Affects PMD Version: 7.0.0-rc3** 

**Rule:UnusedLocalVariable**
<https://docs.pmd-code.org/latest/pmd_rules_java_bestpractices.html#unusedlocalvariable>

**Description:**
When I use an annotation along with a local variable, PMD ignores the usage of the local variable and reports a false positive at line 4.

**Code Sample demonstrating the issue:**

```java
import util.MyAnnotation;
public class ObjectCreator {
  public Supplier<String> create() {
    final Object o = new Object(); // captured by the method ref
    return @MyAnnotation o::toString;
  }
}

```

**Expected outcome: no warning**

PMD reports a violation at line 4, but that's wrong. PMD reports a false positive.

**Running PMD through:** *CLI*

**Comment by adangel** [comment]
I can't get the code example compiled. The annotation at this location (annotating an expression...) doesn't seem to be allowed by java.

@oowekyala Do you know whether this annotation is legal here? I've read here (https://docs.oracle.com/javase/specs/jls/se21/html/jls-9.html#jls-9.7.4) only about declarations... 
PMD still parses this, but considers then "@MyAnnotation o" as a ClassType and then "o" is not recognized as VariableAccess later on.

Is the following line in our grammar correct?
https://github.com/pmd/pmd/blob/76e3b2c99c6621701f461f4fd92740bbea712587/pmd-java/src/main/javacc/Java.jjt#L2076

When removing this, the above code sample is not parseable by PMD either... (then we would behave like javac).

But some other examples then also fail, e.g. https://github.com/pmd/pmd/blob/cd075d24599a74f726a711138ec9b50f790a3c95/pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/ast/ParserCornerCases18.java#L202
This also seems to be invalid java syntax.

And two cases here https://github.com/pmd/pmd/blob/2d33d05d76ea5e6cbf53041305a92b39237bfe16/pmd-java/src/test/kotlin/net/sourceforge/pmd/lang/java/ast/ASTMethodReferenceTest.kt#L198

Are annotations really allowed in expression contexts?

**Comment by oowekyala** [comment]
Type annotations are legal at the start of an expression if they apply to the type at the lhs of a method reference. But that's it, you cannot annotate something that is not a type, like the op is trying to do. That example is illegal code. There is no such thing as an expression annotation. 

AFAIAC our parser behaves correctly, it sees an annotation, and uses that as a cue that the next identifier must be a type.

See
https://github.com/openjdk/jdk/blob/2a406f3ce5e200af9909ce051fdeed0cc059fea0/test/langtools/tools/javac/annotations/typeAnnotations/newlocations/AllLocations.java#L69

**Comment by oowekyala** [comment]
I think we should close this as the input is not valid java code.
